docs/config.md

```diff
@@ -809,6 +809,24 @@ example:
 
     git.push-branch-prefix = "martinvonz/push-"
 
+### Set of private commits
+
+You can configure the set of private commits by setting `git.private-commits` to
+a revset. The value is a revset of commits that Jujutsu will refuse to push. If
+unset, all commits are eligible to be pushed.
```

Comment:
Currently, commits with empty descriptions cannot be pushed even if they are not in `private-commits`, right?

I'm wondering whether or not this is good. The good part of it is that we can have a more informative message.

In any case, should we document it here?

-------------

There is also a similar problem with "immutable_heads()": it is currently very inflexible to configure. There are revisions that are always immutable (`root()`), the `untracked_remotes()` that are a really good idea to keep there, there's `trunk()` which is there by default, and there are any additional heads the user wants to add in there. Currently, to do the latter, the user has to understand all of the former and include them in the config.

I wish we could solve this problem in such a way that there would be a natural way to include commits with no description or author into `private-commits`, without the user having to think about that whenever they want to add something more to their `private-commits`.

Reply:
I see how what I wrote could be confusing. Where I wrote, "If unset, all commits are eligible to be pushed," I should have written, "If unset, no commits are private." The existing checks for authorship, empty description, conflicts, etc. should be unchanged.

Regarding `immutable_heads()`, I do this in my user config:

```toml
"immutable_heads()" = "trunk() | tags() | ... | additional_immutable_heads()"
"additional_immutable_heads()" = "none()"
```

Then in any repo, I can append immutable heads by setting `additional_immutable_heads()`. Maybe jj could do something like this would be baked into the binary? I'm not sure which would be better:

```toml
# jj only uses real_immutable_heads().
# Set immutable_heads() to append to the default revset.
# Set real_immutable_heads() to replace the default revset.
"immutable_heads()"      = "none()"
"real_immutable_heads()" = "trunk() | ... | immutable_heads()"

# jj only uses immutable_heads().
# Set additional_immutable_heads() to append to the default revset.
# Set immutable_heads() to replace the default revset.
"immutable_heads()"            = "trunk() | ... | additional_immutable_heads()"
"additional_immutable_heads()" = "none()"
```

If this would be a good idea for immutable heads, maybe private commits could work similarly? Something to the effect of:

```toml
private-commits = "none()"
real-private-commits = 'description(exact:"") | author(exact:"") | conflicts() | ...`
```

Reply:
I wouldn't add `immutable_heads`/`additional_immutable_heads` abstraction to the default configuration. It can't subtract conditions from the base set, and if user override the base set to achieve that, the additional expression won't be used anymore.

Reply:
Hmm, `builtin_*()` sounds good to me.

Reply:
I also think `builtin_*` is a good idea. It's not perfect -- it's a bit difficult to discover for the user -- but I can't think of anything much better, and perhaps we can make it more discoverable later.

In principle, if needed (which probably it is not yet), we could subdivide the builtin into a union of several other revsets that the user could pick and choose from.

> I see how what I wrote could be confusing. Where I wrote, "If unset, all commits are eligible to be pushed," I should have written, "If unset, no commits are private." The existing checks for authorship, empty description, conflicts, etc. should be unchanged.

Yeah, I think we should update this a bit. Perhaps also mention that the commits without an author can never be pushed.

Also, no problem, I don't think this is problematic, just something that can be improved a bit.